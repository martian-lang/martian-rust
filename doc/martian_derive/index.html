<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="API documentation for the Rust `martian_derive` crate."><meta name="keywords" content="rust, rustlang, rust-lang, martian_derive"><title>martian_derive - Rust</title><link rel="stylesheet" type="text/css" href="../normalize.css"><link rel="stylesheet" type="text/css" href="../rustdoc.css" id="mainThemeStyle"><link rel="stylesheet" type="text/css" href="../light.css"  id="themeStyle"><link rel="stylesheet" type="text/css" href="../dark.css" disabled ><link rel="stylesheet" type="text/css" href="../ayu.css" disabled ><script id="default-settings"></script><script src="../storage.js"></script><script src="../crates.js"></script><noscript><link rel="stylesheet" href="../noscript.css"></noscript><link rel="icon" type="image/svg+xml" href="../favicon.svg">
<link rel="alternate icon" type="image/png" href="../favicon-16x16.png">
<link rel="alternate icon" type="image/png" href="../favicon-32x32.png"><style type="text/css">#crate-search{background-image:url("../down-arrow.svg");}</style></head><body class="rustdoc mod crate"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="sidebar"><div class="sidebar-menu" role="button">&#9776;</div><a href='../martian_derive/index.html'><div class='logo-container rust-logo'><img src='../rust-logo.png' alt='logo'></div></a><h2 class="location">Crate martian_derive</h2><div class="block version"><p>Version 0.24.0</p></div><div class="sidebar-elems"><a id="all-types" href="all.html"><p>See all martian_derive's items</p></a><div class="block items"><ul><li><a href="#macros">Macros</a></li></ul></div><div id="sidebar-vars" data-name="martian_derive" data-ty="mod" data-relpath=""></div><script defer src="sidebar-items.js"></script></div></nav><div class="theme-picker"><button id="theme-picker" aria-label="Pick another theme!" aria-haspopup="menu" title="themes"><img src="../brush.svg" width="18" height="18" alt="Pick another theme!"></button><div id="theme-choices" role="menu"></div></div><nav class="sub"><form class="search-form"><div class="search-container"><div><select id="crate-search"><option value="All crates">All crates</option></select><input class="search-input" name="search" disabled autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"></div><button type="button" id="help-button" title="help">?</button><a id="settings-menu" href="../settings.html" title="settings"><img src="../wheel.svg" width="18" height="18" alt="Change settings"></a></div></form></nav><section id="main" class="content"><h1 class="fqn"><span class="in-band">Crate <a class="mod" href="#">martian_derive</a><button id="copy-path" onclick="copy_path(this)" title="copy path"><img src="../clipboard.svg" width="19" height="18" alt="Copy item import" title="Copy item import to clipboard"></button></span><span class="out-of-band"><span id="render-detail"><a id="toggle-all-docs" href="javascript:void(0)" title="collapse all docs">[<span class="inner">&#x2212;</span>]</a></span><a class="srclink" href="../src/martian_derive/lib.rs.html#1-890" title="goto source code">[src]</a></span></h1><h2 id="macros" class="section-header"><a href="#macros">Macros</a></h2>
<table><tr class="module-item"><td><a class="macro" href="macro.martian_filetype.html" title="martian_derive::martian_filetype macro">martian_filetype</a></td><td class="docblock-short"><p>A macro to define a new struct that implements <code>MartianFileType</code> trait</p>
</td></tr></table><h2 id="attributes" class="section-header"><a href="#attributes">Attribute Macros</a></h2>
<table><tr class="module-item"><td><a class="attr" href="attr.make_mro.html" title="martian_derive::make_mro attr">make_mro</a></td><td class="docblock-short"><p>When this attribute is applied to the <code>MartianMain</code> or <code>MartianStage</code> trait implementation of
a stage struct, it derives the trait <code>MroMaker</code> to the stage struct, which lets you generate
the mro corresponding to the stage.</p>
</td></tr></table><h2 id="derives" class="section-header"><a href="#derives">Derive Macros</a></h2>
<table><tr class="module-item"><td><a class="derive" href="derive.MartianStruct.html" title="martian_derive::MartianStruct derive">MartianStruct</a></td><td class="docblock-short"><p>Structs which are used as associated types in <code>MartianMain</code> or <code>MartianStage</code>
traits need to implement <code>MartianStruct</code>. You can derive it using <code>#[derive(MartianStruct)]</code>
Each field in the struct needs to implement <code>AsMartianBlanketType</code>. This is implemented
for most basic Rust types. You can derive it on custom structs or enum using <code>#[derive(MartianType)]</code>.</p>
</td></tr><tr class="module-item"><td><a class="derive" href="derive.MartianType.html" title="martian_derive::MartianType derive">MartianType</a></td><td class="docblock-short"><p>Custom types which are fields of a <code>MartianStruct</code> need to implement <code>AsMartianBlanketType</code>.
You can derive that trait on an enum or struct using <code>#[derive(MartianType)]</code></p>
</td></tr></table></section><section id="search" class="content hidden"></section><div id="rustdoc-vars" data-root-path="../" data-current-crate="martian_derive" data-search-index-js="../search-index.js" data-search-js="../search.js"></div><script src="../main.js"></script></body></html>