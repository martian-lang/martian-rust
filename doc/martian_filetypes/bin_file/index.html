<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="This module defines a `BincodeFile` and implements `FileTypeIO&lt;T&gt;` and `LazyFileTypeIO&lt;T&gt;` trait for a bincode file. It is essentially a wrapper around the bincode crate. Note that we store additional type info in the file header. Hence the files cannot be directly parsed using the bincode crate functions."><meta name="keywords" content="rust, rustlang, rust-lang, bin_file"><title>martian_filetypes::bin_file - Rust</title><link rel="preload" as="font" type="font/woff2" crossorigin href="../../SourceSerif4-Regular.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../FiraSans-Regular.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../FiraSans-Medium.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../SourceCodePro-Regular.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../SourceSerif4-Bold.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../SourceCodePro-Semibold.ttf.woff2"><link rel="stylesheet" href="../../normalize.css"><link rel="stylesheet" href="../../rustdoc.css" id="mainThemeStyle"><link rel="stylesheet" href="../../ayu.css" disabled><link rel="stylesheet" href="../../dark.css" disabled><link rel="stylesheet" href="../../light.css" id="themeStyle"><script id="default-settings" ></script><script src="../../storage.js"></script><script defer src="../../main.js"></script><noscript><link rel="stylesheet" href="../../noscript.css"></noscript><link rel="alternate icon" type="image/png" href="../../favicon-16x16.png"><link rel="alternate icon" type="image/png" href="../../favicon-32x32.png"><link rel="icon" type="image/svg+xml" href="../../favicon.svg"></head><body class="rustdoc mod"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle">&#9776;</button><a class="sidebar-logo" href="../../martian_filetypes/index.html"><div class="logo-container"><img class="rust-logo" src="../../rust-logo.svg" alt="logo"></div></a><h2></h2></nav><nav class="sidebar"><a class="sidebar-logo" href="../../martian_filetypes/index.html"><div class="logo-container"><img class="rust-logo" src="../../rust-logo.svg" alt="logo"></div></a><h2 class="location"><a href="#">Module bin_file</a></h2><div class="sidebar-elems"><section><ul class="block"><li><a href="#structs">Structs</a></li><li><a href="#types">Type Definitions</a></li></ul></section></div></nav><main><div class="width-limiter"><nav class="sub"><form class="search-form"><span></span><input class="search-input" name="search" autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"><div id="help-button" title="help" tabindex="-1"><a href="../../help.html">?</a></div><div id="settings-menu" tabindex="-1"><a href="../../settings.html" title="settings"><img width="22" height="22" alt="Change settings" src="../../wheel.svg"></a></div></form></nav><section id="main-content" class="content"><div class="main-heading"><h1 class="fqn">Module <a href="../index.html">martian_filetypes</a>::<wbr><a class="mod" href="#">bin_file</a><button id="copy-path" onclick="copy_path(this)" title="Copy item path to clipboard"><img src="../../clipboard.svg" width="19" height="18" alt="Copy item path"></button></h1><span class="out-of-band"><a class="srclink" href="../../src/martian_filetypes/bin_file.rs.html#1-475">source</a> · <a id="toggle-all-docs" href="javascript:void(0)" title="collapse all docs">[<span class="inner">&#x2212;</span>]</a></span></div><details class="rustdoc-toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>This module defines a <code>BincodeFile</code> and implements <code>FileTypeIO&lt;T&gt;</code> and
<code>LazyFileTypeIO&lt;T&gt;</code> trait for a bincode file. It is essentially a wrapper
around the bincode crate. Note that we store additional type info in the file
header. Hence the files cannot be directly parsed using the bincode crate
functions.</p>
<h3 id="compatibility"><a href="#compatibility">Compatibility</a></h3>
<p>Lazily writing items of type T produces a bincode file which <strong>cannot</strong> be read
as <code>Vec&lt;T&gt;</code> using <code>FileTypeIO::read()</code>, instead you need to use the lazy reader
to read items one at a time. The helper fn <code>LazyFileTypeIO::read_all()</code>
can be used to collect all the items in a <code>Vec&lt;T&gt;</code></p>
<h3 id="simple-readwrite-example"><a href="#simple-readwrite-example">Simple read/write example</a></h3>
<p><code>BincodeFile&lt;T&gt;</code> implements <code>FileTypeIO&lt;T&gt;</code> for any serializable type <code>T</code>.</p>

<div class="example-wrap"><pre class="rust rust-example-rendered"><code><span class="kw">use </span>martian_filetypes::{FileTypeRead, FileTypeWrite, bin_file::BincodeFile};
<span class="kw">use </span>martian::Error;
<span class="kw">use </span>serde::{Serialize, Deserialize};

<span class="attribute">#[derive(Debug, PartialEq, Serialize, Deserialize)]
</span><span class="kw">struct </span>Chemistry {
    name: String,
    paired_end: bool,
}

<span class="kw">fn </span>main() -&gt; <span class="prelude-ty">Result</span>&lt;(), Error&gt; {
    <span class="kw">let </span>chem = Chemistry { name: <span class="string">&quot;SCVDJ&quot;</span>.into(), paired_end: <span class="bool-val">true </span>};
    <span class="kw">let </span>bin_file = BincodeFile::from(<span class="string">&quot;bin_example&quot;</span>);
    <span class="comment">// The two functions below are simple wrappers over bincode crate functions
    </span>bin_file.write(<span class="kw-2">&amp;</span>chem)<span class="question-mark">?</span>;
    <span class="kw">let </span>decoded = bin_file.read()<span class="question-mark">?</span>;
    <span class="macro">assert_eq!</span>(chem, decoded);
    <span class="prelude-val">Ok</span>(())
}</code></pre></div>
<h3 id="lazy-readwrite-example"><a href="#lazy-readwrite-example">Lazy read/write example</a></h3>
<p>If the bincode file stores a list of items of type <code>T</code>, then the items can be read
one at a time without reading the whole file into memory. A list of items
of type <code>T</code> can also be incrementally written to a bincode file.
<code>BincodeFile&lt;Vec&lt;T&gt;&gt;</code> implements <code>LazyFileTypeIO&lt;T&gt;</code> for any serializable type <code>T</code>.</p>

<div class="example-wrap"><pre class="rust rust-example-rendered"><code><span class="kw">use </span>martian_filetypes::{FileTypeRead, FileTypeWrite, LazyFileTypeIO, LazyWrite};
<span class="kw">use </span>martian_filetypes::bin_file::{BincodeFile, LazyBincodeReader, LazyBincodeWriter};
<span class="kw">use </span>martian::Error;
<span class="kw">use </span>serde::{Serialize, Deserialize};

<span class="kw">fn </span>main() -&gt; <span class="prelude-ty">Result</span>&lt;(), Error&gt; {
    <span class="kw">let </span>bin_file = BincodeFile::from(<span class="string">&quot;bin_example_lazy&quot;</span>);
    <span class="kw">let </span><span class="kw-2">mut </span>writer = bin_file.lazy_writer()<span class="question-mark">?</span>;
    <span class="comment">// writer implements the trait `LazyWrite&lt;i32&gt;`
    </span><span class="kw">for </span>val <span class="kw">in </span><span class="number">0</span>..<span class="number">10_000i32 </span>{
        writer.write_item(<span class="kw-2">&amp;</span>val)<span class="question-mark">?</span>;
    }
    writer.finish()<span class="question-mark">?</span>; <span class="comment">// The file writing is not completed until finish() is called.

    // We could have collected the vector and invoked write().
    // Both approaches will give you a bincode file which you can lazily read.
    // Note that the bincode files **will not be identical** in the two cases.
    // let vals: Vec&lt;_&gt; = (0..10_000).into_iter().collect()
    // bin_file.write(&amp;vals)?;

    </span><span class="kw">let </span><span class="kw-2">mut </span>reader = bin_file.lazy_reader()<span class="question-mark">?</span>;
    <span class="kw">let </span><span class="kw-2">mut </span>max_val = <span class="number">0</span>;
    <span class="comment">// reader is an `Iterator` over values of type Result&lt;`i32`, Error&gt;
    </span><span class="kw">for </span>(i, val) <span class="kw">in </span>reader.enumerate() {
        <span class="kw">let </span>val: i32 = val<span class="question-mark">?</span>;
        <span class="macro">assert_eq!</span>(i <span class="kw">as </span>i32, val);
        max_val = std::cmp::max(max_val, val);
    }
    <span class="macro">assert_eq!</span>(max_val, <span class="number">9999i32</span>);
    <span class="prelude-val">Ok</span>(())
}</code></pre></div>
</div></details><h2 id="structs" class="small-section-header"><a href="#structs">Structs</a></h2><div class="item-table"><div class="item-row"><div class="item-left module-item"><a class="struct" href="struct.BinaryFormat.html" title="martian_filetypes::bin_file::BinaryFormat struct">BinaryFormat</a></div><div class="item-right docblock-short">A struct that adds typed bincode serialization.</div></div><div class="item-row"><div class="item-left module-item"><a class="struct" href="struct.Bincode.html" title="martian_filetypes::bin_file::Bincode struct">Bincode</a></div></div><div class="item-row"><div class="item-left module-item"><a class="struct" href="struct.LazyBincodeReader.html" title="martian_filetypes::bin_file::LazyBincodeReader struct">LazyBincodeReader</a></div><div class="item-right docblock-short">Iterator over individual items  within a bincode file that
stores a list of items.</div></div><div class="item-row"><div class="item-left module-item"><a class="struct" href="struct.LazyBincodeWriter.html" title="martian_filetypes::bin_file::LazyBincodeWriter struct">LazyBincodeWriter</a></div><div class="item-right docblock-short">Helper struct to write items one by one to a bincode file that
stores a list of items</div></div></div><h2 id="types" class="small-section-header"><a href="#types">Type Definitions</a></h2><div class="item-table"><div class="item-row"><div class="item-left module-item"><a class="type" href="type.BincodeFile.html" title="martian_filetypes::bin_file::BincodeFile type">BincodeFile</a></div></div></div></section></div></main><div id="rustdoc-vars" data-root-path="../../" data-current-crate="martian_filetypes" data-themes="ayu,dark,light" data-resource-suffix="" data-rustdoc-version="1.66.0 (69f9c33d7 2022-12-12)" ></div></body></html>