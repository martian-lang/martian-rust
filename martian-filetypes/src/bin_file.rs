//!
//! This module defines a `BincodeFile` and implements `FileTypeIO<T>` and
//! `LazyFileTypeIO<T>` trait for a bincode file. It is essentially a wrapper
//! around the bincode crate. Note that we store additional type info in the file
//! header. Hence the files cannot be directly parsed using the bincode crate
//! functions.
//!
//! ## Compatibility
//! Lazily writing items of type T produces a bincode file which **cannot** be read
//! as `Vec<T>` using `FileTypeIO::read()`, instead you need to use the lazy reader
//! to read items one at a time. The helper fn `LazyFileTypeIO::read_all()`
//! can be used to collect all the items in a `Vec<T>`
//!
//! ## Simple read/write example
//! `BincodeFile` implements `FileTypeIO<T>` for any type `T` which can be [de]serialized.
//! ```rust
//! use martian_filetypes::{FileTypeIO, bin_file::BincodeFile};
//! use martian::Error;
//! use serde::{Serialize, Deserialize};
//!
//! #[derive(Debug, PartialEq, Serialize, Deserialize)]
//! struct Chemistry {
//!     name: String,
//!     paired_end: bool,
//! }
//!
//! fn main() -> Result<(), Error> {
//!     let chem = Chemistry { name: "SCVDJ".into(), paired_end: true };
//!     let bin_file = BincodeFile::from("bin_example");
//!     // The two functions below are simple wrappers over bincode crate functions
//!     bin_file.write(&chem)?;
//!     let decoded: Chemistry = bin_file.read()?;
//!     assert_eq!(chem, decoded);
//!     # std::fs::remove_file(bin_file)?; // Remove the file (hidden from the doc)
//!     Ok(())
//! }
//! ```
//!
//! ## Lazy read/write example
//! If the bincode file stores a list of items of type `T`, then the items can be read
//! one at a time without reading the whole file into memory. A list of items
//! of type `T` can also be incrementally written to a bincode file.
//! `BincodeFile` implements `LazyFileTypeIO<T>` for any type `T` which can be [de]serialized.
//!
//! ```rust
//! use martian_filetypes::{FileTypeIO, LazyFileTypeIO, LazyWrite};
//! use martian_filetypes::bin_file::{BincodeFile, LazyBincodeReader, LazyBincodeWriter};
//! use martian::Error;
//! use serde::{Serialize, Deserialize};
//!
//! fn main() -> Result<(), Error> {
//!     let bin_file = BincodeFile::from("bin_example_lazy");
//!     let mut writer: LazyBincodeWriter<i32> = bin_file.lazy_writer()?;
//!     // writer implements the trait `LazyWrite<i32>`
//!     for val in 0..10_000i32 {
//!         writer.write_item(&val)?;
//!     }
//!     writer.finish()?; // The file writing is not completed until finish() is called.
//!
//!     // We could have collected the vector and invoked write().
//!     // Both approaches will give you a bincode file which you can lazily read.
//!     // Note that the bincode files **will not be identical** in the two cases.
//!     // let vals: Vec<_> = (0..10_000).into_iter().collect()
//!     // bin_file.write(&vals)?;
//!     
//!     // Type inference engine should figure out the type automatically,
//!     // but it is shown here for illustration.
//!     let mut reader: LazyBincodeReader<i32> = bin_file.lazy_reader()?;
//!     let mut max_val = 0;
//!     // reader is an `Iterator` over values of type Result<`i32`, Error>
//!     for (i, val) in reader.enumerate() {
//!         let val: i32 = val?;
//!         assert_eq!(i as i32, val);
//!         max_val = std::cmp::max(max_val, val);
//!     }
//!     assert_eq!(max_val, 9999i32);
//!     # std::fs::remove_file(bin_file)?; // Remove the file (hidden from the doc)
//!     Ok(())
//! }
//! ```

use crate::{FileStorage, FileTypeIO, LazyAgents, LazyRead, LazyWrite};
use failure::format_err;
use martian::{Error, MartianFileType};
use martian_derive::martian_filetype;
use serde::de::DeserializeOwned;
use serde::{Deserialize, Serialize};
use std::any::{type_name, Any};
use std::fmt::Debug;
use std::fs::File;
use std::io::{self, BufReader, BufWriter, Read, Write};
use std::iter::Iterator;
use std::marker::PhantomData;

martian_filetype! {Bincode, "bincode"}
impl<T> FileStorage<T> for Bincode where T: Serialize + DeserializeOwned {}
pub type BincodeFile = BinaryFormat<Bincode>;

crate::martian_filetype_inner! {
    /// Binary format
    pub struct BinaryFormat, "bincode"
}

impl<F, T> FileStorage<T> for BinaryFormat<F> where F: MartianFileType + FileStorage<T> {}

/// Any type `T` that can be deserialized implements `load()` from a `BincodeFile`
/// TODO: Include the TypeId here?
impl<T, F> FileTypeIO<T> for BinaryFormat<F>
where
    T: Any + Serialize + DeserializeOwned,
    F: FileStorage<T> + Debug,
{
    fn read_from<R: Read>(mut reader: R) -> Result<T, Error> {
        let actual_type: String = bincode::deserialize_from(&mut reader)?;

        if actual_type == type_name::<T>() {
            Ok(bincode::deserialize_from(&mut reader)?)
        } else if actual_type == type_name::<LazyMarker<T>>() {
            Err(format_err!(
                "Lazily written bincode files cannot be read using FileTypeIO::read(). Use LazyFileTypeIO::read_all() instead."
            ))
        } else {
            Err(format_err!(
                "Data type {} does not match expected type {} or {}",
                actual_type,
                type_name::<T>(),
                type_name::<LazyMarker<T>>()
            ))
        }
    }

    fn write_into<W: Write>(mut writer: W, input: &T) -> Result<(), Error> {
        let type_hash = type_name::<T>();
        bincode::serialize_into(&mut writer, &type_hash)?;
        bincode::serialize_into(&mut writer, &input)?;
        Ok(())
    }
}

enum FileMode {
    Vec(usize),
    Lazy,
}

/// Iterator over individual items  within a bincode file that
/// stores a list of items.
pub struct LazyBincodeReader<T, F = Bincode, R = BufReader<File>>
where
    F: MartianFileType + FileStorage<Vec<T>>,
    R: Read,
    T: Any + DeserializeOwned,
{
    reader: R,
    mode: FileMode,
    processed_items: usize,
    phantom_f: PhantomData<F>,
    phantom_t: PhantomData<T>,
}

impl<T, F, R> LazyRead<T, R> for LazyBincodeReader<T, F, R>
where
    F: MartianFileType + FileStorage<Vec<T>>,
    R: Read,
    T: Any + Serialize + DeserializeOwned,
{
    type FileType = BinaryFormat<F>;
    fn with_reader(mut reader: R) -> Result<Self, Error> {
        let actual_type: String = bincode::deserialize_from(&mut reader)?;
        let mode = if actual_type == type_name::<Vec<T>>() {
            let total_items: usize = bincode::deserialize_from(&mut reader)?;
            FileMode::Vec(total_items)
        } else if actual_type == type_name::<LazyMarker<Vec<T>>>() {
            FileMode::Lazy
        } else {
            return Err(format_err!(
                "Data type {} does not match expected type {} or {}",
                actual_type,
                type_name::<Vec<T>>(),
                type_name::<LazyMarker<Vec<T>>>()
            ));
        };

        Ok(LazyBincodeReader {
            reader,
            mode,
            processed_items: 0,
            phantom_f: PhantomData,
            phantom_t: PhantomData,
        })
    }
}

impl<T, F, R> Iterator for LazyBincodeReader<T, F, R>
where
    F: MartianFileType + FileStorage<Vec<T>>,
    R: Read,
    T: Any + DeserializeOwned,
{
    type Item = Result<T, Error>;
    fn next(&mut self) -> Option<Self::Item> {
        match self.mode {
            FileMode::Vec(total_items) => {
                if self.processed_items < total_items {
                    self.processed_items += 1;
                    match bincode::deserialize_from(&mut self.reader) {
                        Ok(t) => Some(Ok(t)),
                        Err(e) => Some(Err(Error::from(e))),
                    }
                } else {
                    None
                }
            }
            FileMode::Lazy => match try_bincode_deserialize(&mut self.reader) {
                DeserializeResult::Item(t) => Some(Ok(t)),
                DeserializeResult::Eof => None,
                DeserializeResult::ReadError(e) => Some(Err(e)),
            },
        }
    }
}

enum DeserializeResult<T, E> {
    Item(T),
    ReadError(E),
    Eof,
}

fn try_bincode_deserialize<T, R>(reader: &mut R) -> DeserializeResult<T, Error>
where
    R: Read,
    T: DeserializeOwned,
{
    match bincode::deserialize_from(reader) {
        Ok(t) => DeserializeResult::Item(t),
        Err(e) => {
            match *e.as_ref() {
                bincode::ErrorKind::Io(ref io_e) => {
                    match io_e.kind() {
                        io::ErrorKind::UnexpectedEof => DeserializeResult::Eof, // We are at the end of the stream
                        _ => DeserializeResult::ReadError(Error::from(e)),
                    }
                }
                _ => DeserializeResult::ReadError(Error::from(e)),
            }
        }
    }
}

// When a list of items of type T is lazily written,
// the typeid we store is that of LazyMarker<Vec<T>>
struct LazyMarker<T>(PhantomData<T>);

/// Helper struct to write items one by one to a bincode file that
/// stores a list of items
pub struct LazyBincodeWriter<T, F = Bincode, W = BufWriter<File>>
where
    F: MartianFileType + FileStorage<Vec<T>>,
    W: Write,
    T: Any + Serialize,
{
    writer: W,
    phantom_f: PhantomData<F>,
    phantom_t: PhantomData<T>,
    processed_items: usize,
}

impl<T, F, W> LazyWrite<T, W> for LazyBincodeWriter<T, F, W>
where
    F: MartianFileType + FileStorage<Vec<T>>,
    W: Write,
    T: Any + Serialize,
{
    type FileType = BinaryFormat<F>;
    fn with_writer(mut writer: W) -> Result<Self, Error> {
        let type_hash = type_name::<LazyMarker<Vec<T>>>(); // The file stores Vec<T>, not T
        bincode::serialize_into(&mut writer, &type_hash)?;
        Ok(LazyBincodeWriter {
            writer,
            phantom_f: PhantomData,
            phantom_t: PhantomData,
            processed_items: 0,
        })
    }
    fn write_item(&mut self, item: &T) -> Result<(), Error> {
        self.processed_items += 1;
        bincode::serialize_into(&mut self.writer, &item)?;
        Ok(())
    }
    fn finish(self) -> Result<W, Error> {
        Ok(self.writer)
    }
}

impl<T, F, W, R> LazyAgents<T, W, R> for BinaryFormat<F>
where
    F: MartianFileType + FileStorage<Vec<T>>,
    R: Read,
    W: Write,
    T: Any + Serialize + DeserializeOwned,
{
    type LazyWriter = LazyBincodeWriter<T, F, W>;
    type LazyReader = LazyBincodeReader<T, F, R>;
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::LazyFileTypeIO;
    use martian::MartianFileType;
    use proptest::arbitrary::any;
    use proptest::collection::vec;
    use proptest::{prop_assert, proptest};

    #[derive(Debug, Serialize, Deserialize, PartialEq, Clone)]
    enum Bar {
        Variant,
    }
    #[derive(Debug, Serialize, Deserialize, PartialEq, Clone)]
    struct Foo {
        v1: u32,
        v2: String,
        v3: Bar,
    }

    proptest! {
        #[test]
        fn prop_test_bincode_file_u8(
            ref seq in vec(any::<u8>(), 0usize..1000usize),
        ) {
            prop_assert!(crate::round_trip_check::<BincodeFile, _>(seq).unwrap());
            prop_assert!(crate::lazy_round_trip_check::<BincodeFile, _>(seq, false).unwrap());
        }
        #[test]
        fn prop_test_bincode_file_bool(
            ref seq in vec(any::<bool>(), 0usize..1000usize),
        ) {
            prop_assert!(crate::round_trip_check::<BincodeFile, _>(seq).unwrap());
            prop_assert!(crate::lazy_round_trip_check::<BincodeFile, _>(seq, false).unwrap());
        }
        #[test]
        fn prop_test_bincode_file_vec_string(
            ref seq in vec(any::<String>(), 0usize..20usize),
        ) {
            prop_assert!(crate::round_trip_check::<BincodeFile, _>(seq).unwrap());
            prop_assert!(crate::lazy_round_trip_check::<BincodeFile, _>(seq, false).unwrap());
        }
        #[test]
        fn prop_test_bincode_file_string(
            ref seq in any::<String>(),
        ) {
            prop_assert!(crate::round_trip_check::<BincodeFile, _>(seq).unwrap());
        }
        #[test]
        fn prop_test_bincode_file_struct(
            ref v1 in any::<u32>(),
            ref v2 in any::<String>(),
        ) {
            let foo = Foo {v1: *v1, v2: v2.clone(), v3: Bar::Variant};
            prop_assert!(crate::round_trip_check::<BincodeFile, _>(&foo).unwrap());

            let input = vec![foo.clone(); 20];
            prop_assert!(crate::round_trip_check::<BincodeFile, _>(&input).unwrap());
            prop_assert!(crate::lazy_round_trip_check::<BincodeFile, _>(&input, false).unwrap());

            let input = vec![vec![foo.clone(); 2]; 4];
            prop_assert!(crate::round_trip_check::<BincodeFile, _>(&input).unwrap());
            prop_assert!(crate::lazy_round_trip_check::<BincodeFile, _>(&input, false).unwrap());

        }
    }

    #[test]
    fn test_bincode_simple_roundtrip() -> Result<(), Error> {
        #[derive(Serialize, Deserialize, PartialEq)]
        struct Foo {
            name: String,
            id: usize,
        }
        let foo = Foo {
            name: "Bar".into(),
            id: 100,
        };
        assert!(crate::round_trip_check::<BincodeFile, _>(&foo)?);
        Ok(())
    }

    #[test]
    fn test_bincode_inconsistent_type() -> Result<(), Error> {
        let dir = tempfile::tempdir()?;

        let fn1 = BincodeFile::new(dir.path(), "test1");
        let fn2 = BincodeFile::new(dir.path(), "test2");

        let v1 = vec![1u8, 2u8, 3u8];
        let v2 = vec![1u64, 2u64, 3u64];
        fn1.write(&v1)?;
        fn2.write(&v2)?;

        let _: Vec<u8> = fn1.read()?;
        let _: Vec<u64> = fn2.read()?;

        assert!(FileTypeIO::<Vec<u64>>::read(&fn1).is_err());
        assert!(FileTypeIO::<u8>::read(&fn1).is_err());
        assert!(FileTypeIO::<String>::read(&fn1).is_err());

        assert!(FileTypeIO::<Vec<u8>>::read(&fn2).is_err());
        assert!(FileTypeIO::<u8>::read(&fn2).is_err());
        assert!(FileTypeIO::<String>::read(&fn2).is_err());

        Ok(())
    }

    #[test]
    fn test_lazy_read() -> Result<(), Error> {
        let values: Vec<u16> = (0..100).into_iter().collect();
        let dir = tempfile::tempdir()?;
        let bin_file = BincodeFile::new(dir.path(), "my_file");
        bin_file.write(&values)?;

        let inc_reader = bin_file.lazy_reader()?;
        for (i, v) in inc_reader.enumerate() {
            let v: u16 = v?;
            assert_eq!(i as u16, v);
        }

        let max_val = bin_file.lazy_reader()?.map(|x| x.unwrap()).max();
        assert_eq!(max_val, Some(99u16));

        // Invalid Type
        assert!(LazyBincodeReader::<u8>::with_reader(bin_file.buf_reader()?).is_err());

        Ok(())
    }

    #[test]
    fn test_bincode_lazy_write_and_read() -> Result<(), Error> {
        let dir = tempfile::tempdir()?;
        let bin_file = BincodeFile::new(dir.path(), "my_file");
        let mut writer = bin_file.lazy_writer()?;
        for i in 0..10i32 {
            writer.write_item(&i)?;
        }
        writer.finish()?;

        let r: Result<Vec<i32>, Error> = bin_file.read();
        assert!(r.is_err());
        Ok(())
    }
}
