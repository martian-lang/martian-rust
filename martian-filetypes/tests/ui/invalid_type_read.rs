use anyhow::Error;
use martian_derive::martian_filetype;
use martian_filetypes::bin_file::BinaryFormat;
use martian_filetypes::json_file::JsonFormat;
use martian_filetypes::{FileTypeRead, FileTypeWrite};
use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize, Deserialize)]
struct Feature {
    id: usize,
}

#[derive(Debug, Serialize, Deserialize)]
struct Creature {
    id: usize,
}

martian_filetype! {FeatureFile, "feat"}

fn main() -> Result<(), Error> {
    let feature = Feature { id: 5 };
    let creature = Creature { id: 10 };
    // Json Format
    {
        let feat_file: JsonFormat<FeatureFile, Feature> = JsonFormat::from("feature");
        feat_file.write(&feature)?;
        let new_feature: Creature = feat_file.read()?; // Compiler error
        std::fs::remove_file(feat_file)?;
    }
    // Binary Format
    {
        let feat_file: BinaryFormat<FeatureFile, Feature> = BinaryFormat::from("feature");
        feat_file.write(&feature)?;
        let new_feature: Creature = feat_file.read()?; // Compiler error
        std::fs::remove_file(feat_file)?;
    }
    Ok(())
}
