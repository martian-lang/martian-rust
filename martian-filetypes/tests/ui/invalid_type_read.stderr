error[E0308]: `?` operator has incompatible types
  --> tests/ui/invalid_type_read.rs:27:37
   |
27 |         let new_feature: Creature = feat_file.read()?; // Compiler error
   |                                     ^^^^^^^^^^^^^^^^^ expected struct `Creature`, found struct `Feature`
   |
   = note: `?` operator cannot convert from `Feature` to `Creature`

error[E0308]: `?` operator has incompatible types
  --> tests/ui/invalid_type_read.rs:34:37
   |
34 |         let new_feature: Creature = feat_file.read()?; // Compiler error
   |                                     ^^^^^^^^^^^^^^^^^ expected struct `Creature`, found struct `Feature`
   |
   = note: `?` operator cannot convert from `Feature` to `Creature`
